core/src/main/java/com/netflix/msl/msg/MessageHeader.java

```diff
@@ -268,6 +270,20 @@ public MessageHeader(final MslContext ctx, final EntityAuthenticationData entity
         }
         if (!encrypted && headerData.userAuthData != null)
             throw new MslInternalException("User authentication data cannot be included if the message is not encrypted.");
+        
+        // Older MSL stacks expect the sender if a master token is being used.
+        //
+        // If the local entity does not know its entity identity, then use the
+        // empty string. This will work except for the case where the old MSL
+        // stack is receiving a message for which it is also the issuer of the
+        // master token. That scenario will continue to fail.
```

Comment:
How do these clients recover from the failures?

Reply:
Only servers (trusted services or peer-to-peer) receive messages containing master tokens it issued. To my knowledge there are no such entities running old MSL stacks.

With the newer support for entities that do not know their own identity, it is vital that servers do not look for a sender field at all.